linera-core/src/client.rs

```diff
@@ -1294,17 +1316,215 @@ where
+                .try_process_certificates(name, &mut node, chain_id, certificates, notifications)
+                .await
+                .is_none()
+        {
+            return Ok(());
+        };
+        if let Some(proposal) = info.manager.requested_proposed {
+            if proposal.content.block.chain_id == chain_id {
+                let owner = proposal.owner;
+                loop {
+                    let result = self
+                        .client
+                        .local_node
+                        .handle_block_proposal(*proposal.clone())
+                        .await;
+                    if let Err(original_err) = &result {
+                        if let LocalNodeError::WorkerError(WorkerError::ChainError(chain_error)) =
+                            original_err
+                        {
+                            if let ChainError::ExecutionError(
+                                ExecutionError::SystemError(
+                                    SystemExecutionError::BlobNotFoundOnRead(blob_id),
+                                ),
+                                _,
+                            ) = &**chain_error
+                            {
+                                self.update_local_node_with_blob(*blob_id).await?;
+                                continue;
+                            }
+                        }
+                        tracing::warn!("Skipping proposal from {}: {}", owner, original_err);
+                    }
+                    break;
+                }
+            }
+        }
+        if let Some(cert) = info.manager.requested_locked {
+            if cert.value().is_validated() && cert.value().chain_id() == chain_id {
+                let hash = cert.hash();
+                if let Err(error) = self
+                    .client
+                    .local_node
+                    .handle_certificate(*cert.clone(), vec![], vec![], notifications)
+                    .await
+                {
+                    tracing::warn!("Skipping certificate {}: {}", hash, error);
+                    return Err(error.into());
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn update_local_node_with_blob(&self, blob_id: BlobId) -> Result<(), ChainClientError> {
+        let certificate = self.get_blobs_confirmed_certificate(blob_id).await?;
+        // This will download all ancestors of the certificate and process all of them locally.
+        self.receive_certificate(certificate).await?;
+        Ok(())
+    }
+
+    async fn get_blobs_confirmed_certificate(
+        &self,
+        blob_id: BlobId,
+    ) -> Result<Certificate, ChainClientError> {
+        // Even after synching this chain with the validators, we still don't have the blob we're trying to read.
+        // Which means it was published by a different chain, so we need to update this chain with it.
+        let validators = self.validator_nodes().await?;
```

Comment:
Isn't all of this already happening in a single `try_synchronize_chain_state_from` call, with a single validator? And that validator is the one who gave us the block with the missing blob? Then it should be sufficient to ask that one validator for the blob and certificate, rather than all of them.